File: code_snippets/ts/src/get_started/tour.ts
import * as restate from "@restatedev/restate-sdk";

const router = restate.service({
  name: "MyService",
  handlers: {
    greet: async (ctx: restate.Context, name: string) => {
      const ticketId = "";

      // <start_sleep>
      await ctx.sleep({ minutes: 15 });
      // <end_sleep>

      // <start_sleep_and_send>
      await ctx.sleep({ minutes: 15 });
      ctx.objectSendClient(TicketObject, ticketId).unreserve();
      // <end_sleep_and_send>
    },
  },
});

const checkoutService = restate.service({
  name: "MyService",
  handlers: {
    // <start_uuid>
    async handle(
      ctx: restate.Context,
      request: { userId: string; tickets: string[] }
    ) {
      // !mark(1:3)
      const idempotencyKey = ctx.rand.uuidv4();
      console.info("My idempotency key: " + idempotencyKey);
      throw new Error("Something happened!");

      return true;
    },
    // <end_uuid>
  },
});

const secondCheckoutService = restate.service({
  name: "MyService",
  handlers: {
    // <start_checkout>
    async handle(
      ctx: restate.Context,
      request: { userId: string; tickets: string[] }
    ) {
      // !mark
      const totalPrice = request.tickets.length * 40;

      const idempotencyKey = ctx.rand.uuidv4();
      // !mark(1:3)
      const success = await ctx.run(() =>
        PaymentClient.get().call(idempotencyKey, totalPrice)
      );

      return success;
    },
    // <end_checkout>
  },
});
export const ticketObject = restate.object({
  name: "TicketObject",
  handlers: {
    async unreserve(ctx: restate.ObjectContext) {
      return true;
    },
  },
});

const TicketObject: typeof ticketObject = { name: "TicketObject" };

class PaymentClient {
  static get() {
    return new PaymentClient();
  }

  async call(idempotencyKey: string, totalPrice: number) {
    return true;
  }
}

File: code_snippets/rust/src/get_started/tour.rs
#![allow(dead_code)]

use restate_sdk::prelude::*;
use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use std::time::Duration;
use tracing::info;

#[restate_sdk::object]
pub(crate) trait TicketObject {
    async fn unreserve() -> Result<(), HandlerError>;
}

async fn add_ticket(ctx: ObjectContext<'_>) -> Result<(), HandlerError> {
    // <start_sleep>
    ctx.sleep(Duration::from_millis(15 * 60 * 1000)).await?;
    // <end_sleep>

    // <start_sleep_and_send>
    ctx.sleep(Duration::from_millis(15 * 60 * 1000)).await?;
    ctx.object_client::<TicketObjectClient>(ctx.key())
        .unreserve()
        .send();
    // <end_sleep_and_send>
    Ok(())
}

#[derive(Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct CheckoutRequest {
    pub(crate) user_id: String,
    pub(crate) tickets: HashSet<String>,
}

#[restate_sdk::service]
pub(crate) trait CheckoutService {
    async fn handle(request: Json<CheckoutRequest>) -> Result<bool, HandlerError>;
}

pub struct CheckoutServiceImpl;

#[allow(unused_variables)]
impl CheckoutService for CheckoutServiceImpl {
    // <start_uuid>
    async fn handle(
        &self,
        mut ctx: Context<'_>,
        Json(request): Json<CheckoutRequest>,
    ) -> Result<bool, HandlerError> {
        let idempotency_key = ctx.rand_uuid().to_string();
        info!("idempotent key: {}", idempotency_key);

        Err(HandlerError::from("Something happened!"))
    }
    // <end_uuid>
}

pub struct OtherCheckoutService;

#[allow(unused_variables)]
impl CheckoutService for OtherCheckoutService {
    // <start_checkout>
    async fn handle(
        &self,
        mut ctx: Context<'_>,
        Json(CheckoutRequest { user_id, tickets }): Json<CheckoutRequest>,
    ) -> Result<bool, HandlerError> {
        // !mark
        let total_price = tickets.len() as f64 * 40.0;

        let idempotency_key = ctx.rand_uuid().to_string();

        // !mark(1:4)
        let pay_client = PaymentClient::new();
        let success = ctx
            .run(|| pay_client.call(&idempotency_key, total_price))
            .await?;

        Ok(success)
    }
    // <end_checkout>
}

pub struct PaymentClient {}

impl PaymentClient {
    pub fn new() -> Self {
        PaymentClient {}
    }

    #[allow(unused_variables)]
    pub async fn call(&self, idempotency_key: &str, amount: f64) -> Result<bool, HandlerError> {
        Ok(true)
    }
}
